internal/credentials/kubernetes/acr/workload_identity.go

```diff
@@ -0,0 +1,330 @@
+
+	// The following behaviors are overridable for testing purposes:
+	getProjectFn    func(ctx context.Context, project string) (*v1alpha1.Project, error)
+	exchangeTokenFn func(ctx context.Context, token, clientID, oAuthEndpoint, scope string) (string, error)
+	fetchSAFn       func(ctx context.Context, project string, audiences []string) (string, error)
+	fetchACRTokenFn func(endpoint, tokenType string, data url.Values) (string, error)
+}
+
+// NewWorkloadIdentityCredentialHelper returns an implementation
+// credentials.Helper that utilizes a cache to avoid unnecessary calls to Azure.
+func NewWorkloadIdentityCredentialHelper(ctx context.Context, kargoClient client.Client) credentials.Helper {
+	logger := logging.LoggerFromContext(ctx)
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+
+	env, err := getAzureEnvironment()
+	if err != nil {
+		logger.Info("Azure environment not set; Azure Workload Identity integration is disabled")
+		return nil
+	}
+	logger.Info("Azure Workload Identity integration is enabled")
+
+	w := &workloadIdentityCredentialHelper{
+		tenantID:    tenantID,
+		clientID:    clientID,
+		kargoClient: kargoClient,
+		azureEnv:    &env,
+		tokenCache: cache.New(
+			// Access tokens live for three hours. We'll hang on to them for 40 minutes.
+			40*time.Minute, // Default ttl for each entry
+			time.Hour,      // Cleanup interval
+		),
+	}
+
+	w.fetchSAFn = w.fetchSAToken
+	w.getProjectFn = w.getProject
+	w.fetchACRTokenFn = fetchACRToken
+	w.exchangeTokenFn = exchangeForEntraIDToken
+
+	return w.getCredentials
+}
+
+func (w *workloadIdentityCredentialHelper) getCredentials(
+	ctx context.Context,
+	project string,
+	credType credentials.Type,
+	repoURL string,
+	_ *corev1.Secret) (*credentials.Credentials, error) {
+	logger := logging.LoggerFromContext(ctx)
+
+	// Workload Identity isn't set up for this controller
+	if (credType != credentials.TypeImage && credType != credentials.TypeHelm) || w.azureEnv == nil {
+		// This helper can't handle this
+		return nil, nil
+	}
+
+	if credType == credentials.TypeHelm && !strings.HasPrefix(repoURL, "oci://") {
+		// Only OCI Helm repos are supported in ACR
+		return nil, nil
+	}
+
+	// TODO: add regex to verify that the URL is an Azure CR URL.
+
+	cacheKey := w.tokenCacheKey(repoURL, project)
+
+	if entry, exists := w.tokenCache.Get(cacheKey); exists {
+		return &credentials.Credentials{
+			Username: accessTokenUsername,
+			Password: entry.(string), // nolint: forcetypeassert
+		}, nil
+	}
+
+	proj, err := w.getProjectFn(ctx, project)
+	if err != nil {
+		return nil, err
+	}
+
+	clientID, ok := proj.GetAnnotations()[AnnotationClientID]
+	if !ok {
+		return nil, fmt.Errorf("project is missing annotation: %s", AnnotationClientID)
+	}
+	tenantID, ok := proj.GetAnnotations()[AnnotationTenantID]
+	if !ok {
+		return nil, fmt.Errorf("project is missing annotation: %s", AnnotationTenantID)
+	}
+
+	repoHost, err := url.Parse(repoURL)
```

Comment:
`repoHost` isn't the host. It's a `url.URL`, so I would perhaps reconsider the naming of this var to prevent confusion.

Reply:
Renamed it to `repoURL`, and renamed the string which it's constructed from to `repo`. I don't really like the string name but can't think of anything better atm...

Reply:
Personally, in a case like this where I have a string I will parse into some other type, I add a suffix to the string var. So here, `repoURLStr` and `repoURL`, but that's, as I said, a personal convention. Feel free to use it or not.